File: simulation.py
import random
import math

# Bandit Arms (True Means)
means = {'A': 0.8, 'B': 0.7, 'C': 0.5}
arms = list(means.keys())
budget = 10000

# Simulation Parameters
num_runs = 50 

def run_simulation():
    results = {}

    # 1. A/B Testing
    def ab_test():
        exploration_budget = 2000
        decision_budget = 8000
        # Exploration
        trials_per_arm = exploration_budget // 3
        rewards = {arm: 0 for arm in arms}
        for arm in arms:
            for _ in range(trials_per_arm):
                if random.random() < means[arm]:
                    rewards[arm] += 1
        
        counts = {arm: trials_per_arm for arm in arms}
        winner = max(arms, key=lambda a: rewards[a] / (counts[a] or 1))
        
        # Decision
        decision_reward = sum(1 for _ in range(decision_budget) if random.random() < means[winner])
        total_reward = sum(rewards.values()) + decision_reward
        return total_reward

    # 2. ε-Greedy (ε=0.1)
    def epsilon_greedy(epsilon=0.1):
        rewards_sum = {arm: 0 for arm in arms}
        counts = {arm: 0 for arm in arms}
        total_reward = 0
        for _ in range(budget):
            if random.random() < epsilon:
                arm = random.choice(arms)
            else:
                q_values = {a: (rewards_sum[a] / counts[a]) if counts[a] > 0 else 1.0 for a in arms}
                arm = max(arms, key=lambda a: q_values[a])
            
            reward = 1 if random.random() < means[arm] else 0
            rewards_sum[arm] += reward
            counts[arm] += 1
            total_reward += reward
        return total_reward

    # 3. Optimistic Initial Values (OIV=2.0)
    def optimistic_initial_values(initial_q=2.0):
        rewards_sum = {arm: 0 for arm in arms}
        counts = {arm: 0 for arm in arms}
        total_reward = 0
        q_values = {arm: initial_q for arm in arms}
        for _ in range(budget):
            arm = max(arms, key=lambda a: q_values[a])
            reward = 1 if random.random() < means[arm] else 0
            rewards_sum[arm] += reward
            counts[arm] += 1
            q_values[arm] = (rewards_sum[arm] + initial_q) / (counts[arm] + 1) # simple update
            total_reward += reward
        return total_reward

    # 4. Softmax (τ=0.1)
    def softmax(tau=0.1):
        rewards_sum = {arm: 0 for arm in arms}
        counts = {arm: 1 for arm in arms} 
        total_reward = 0
        for _ in range(budget):
            q_values = [rewards_sum[a] / counts[a] for a in arms]
            exp_q = [math.exp(q / tau) for q in q_values]
            sum_exp = sum(exp_q)
            probs = [e / sum_exp for e in exp_q]
            
            r = random.random()
            acc = 0
            arm = arms[-1]
            for i, p in enumerate(probs):
                acc += p
                if r < acc:
                    arm = arms[i]
                    break
            
            reward = 1 if random.random() < means[arm] else 0
            rewards_sum[arm] += reward
            counts[arm] += 1
            total_reward += reward
        return total_reward

    # 5. UCB1
    def ucb():
        rewards_sum = {arm: 0 for arm in arms}
        counts = {arm: 0 for arm in arms}
        total_reward = 0
        for arm in arms:
            reward = 1 if random.random() < means[arm] else 0
            rewards_sum[arm] += reward
            counts[arm] += 1
            total_reward += reward
            
        for t in range(len(arms), budget):
            q_values = {a: (rewards_sum[a] / counts[a]) + math.sqrt(2 * math.log(t) / counts[a]) for a in arms}
            arm = max(arms, key=lambda a: q_values[a])
            reward = 1 if random.random() < means[arm] else 0
            rewards_sum[arm] += reward
            counts[arm] += 1
            total_reward += reward
        return total_reward

    # Thompson Sampling simplified using Beta sampling via random.betavariate
    def thompson():
        successes = {arm: 1 for arm in arms}
        failures = {arm: 1 for arm in arms}
        total_reward = 0
        for _ in range(budget):
            samples = {a: random.betavariate(successes[a], failures[a]) for a in arms}
            arm = max(arms, key=lambda a: samples[a])
            reward = 1 if random.random() < means[arm] else 0
            successes[arm] += reward
            failures[arm] += (1 - reward)
            total_reward += reward
        return total_reward

    sims = {
        'A/B Test': ab_test,
        'ε-Greedy': epsilon_greedy,
        'Optimistic': optimistic_initial_values,
        'Softmax': softmax,
        'UCB': ucb,
        'Thompson': thompson
    }

    results = {name: [] for name in sims}
    for _ in range(num_runs):
        for name, func in sims.items():
            results[name].append(func())

    summary = {}
    optimal_reward = 8000 # 10000 * 0.8
    for name, r_list in results.items():
        avg_reward = sum(r_list) / len(r_list)
        summary[name] = {
            'Avg Reward': round(avg_reward, 2),
            'Regret': round(optimal_reward - avg_reward, 2)
        }
    
    return summary

if __name__ == "__main__":
    print(run_simulation())

File: simulation.js
const math = Math;

const means = { A: 0.8, B: 0.7, C: 0.5 };
const arms = ['A', 'B', 'C'];
const budget = 10000;
const numRuns = 10;

function randomBeta(alpha, beta) {
    // Basic approximation of Beta sampling using Gamma sampling
    // Or just use a simple rejection sampler for small alpha, beta
    // For Thompson sampling in class, we can just use a simpler model or assume these libraries exist.
    // Let's implement a simple approximation for Beta since we only have successes/failures.
    function gamma(alpha) {
        let res = 0;
        for (let i = 0; i < alpha; i++) {
            res += -math.log(math.random());
        }
        return res;
    }
    const x = gamma(alpha);
    const y = gamma(beta);
    return x / (x + y);
}

function simulate() {
    const sims = {
        'A/B Test': () => {
            const expBudget = 2000;
            const decBudget = 8000;
            const trialsPerArm = math.floor(expBudget / 3);
            let rewards = { A: 0, B: 0, C: 0 };
            for (let arm of arms) {
                for (let i = 0; i < trialsPerArm; i++) {
                    if (math.random() < means[arm]) rewards[arm]++;
                }
            }
            const winner = arms.reduce((a, b) => (rewards[a] / trialsPerArm >= rewards[b] / trialsPerArm ? a : b));
            let decReward = 0;
            for (let i = 0; i < decBudget; i++) {
                if (math.random() < means[winner]) decReward++;
            }
            return Object.values(rewards).reduce((s, r) => s + r, 0) + decReward;
        },
        'ε-Greedy': () => {
            const epsilon = 0.1;
            let rewardsSum = { A: 0, B: 0, C: 0 };
            let counts = { A: 0, B: 0, C: 0 };
            let total = 0;
            for (let t = 0; t < budget; t++) {
                let arm;
                if (math.random() < epsilon) {
                    arm = arms[math.floor(math.random() * arms.length)];
                } else {
                    arm = arms.reduce((a, b) => {
                        let qa = counts[a] > 0 ? rewardsSum[a] / counts[a] : 1.0;
                        let qb = counts[b] > 0 ? rewardsSum[b] / counts[b] : 1.0;
                        return qa >= qb ? a : b;
                    });
                }
                const reward = math.random() < means[arm] ? 1 : 0;
                rewardsSum[arm] += reward;
                counts[arm]++;
                total += reward;
            }
            return total;
        },
        'Optimistic': () => {
            const initialQ = 2.0;
            let rewardsSum = { A: 0, B: 0, C: 0 };
            let counts = { A: 0, B: 0, C: 0 };
            let total = 0;
            let qValues = { A: initialQ, B: initialQ, C: initialQ };
            for (let t = 0; t < budget; t++) {
                const arm = arms.reduce((a, b) => (qValues[a] >= qValues[b] ? a : b));
                const reward = math.random() < means[arm] ? 1 : 0;
                rewardsSum[arm] += reward;
                counts[arm]++;
                qValues[arm] = (rewardsSum[arm] + initialQ) / (counts[arm] + 1);
                total += reward;
            }
            return total;
        },
        'Softmax': () => {
            const tau = 0.1;
            let rewardsSum = { A: 0, B: 0, C: 0 };
            let counts = { A: 1, B: 1, C: 1 };
            let total = 0;
            for (let t = 0; t < budget; t++) {
                let expQs = arms.map(a => math.exp((rewardsSum[a] / counts[a]) / tau));
                let sumExp = expQs.reduce((s, e) => s + e, 0);
                let probs = expQs.map(e => e / sumExp);
                let r = math.random();
                let acc = 0;
                let arm = arms[arms.length - 1];
                for (let i = 0; i < probs.length; i++) {
                    acc += probs[i];
                    if (r < acc) {
                        arm = arms[i];
                        break;
                    }
                }
                const reward = math.random() < means[arm] ? 1 : 0;
                rewardsSum[arm] += reward;
                counts[arm]++;
                total += reward;
            }
            return total;
        },
        'UCB': () => {
            let rewardsSum = { A: 0, B: 0, C: 0 };
            let counts = { A: 0, B: 0, C: 0 };
            let total = 0;
            for (let arm of arms) {
                const reward = math.random() < means[arm] ? 1 : 0;
                rewardsSum[arm] += reward;
                counts[arm]++;
                total += reward;
            }
            for (let t = arms.length; t < budget; t++) {
                const arm = arms.reduce((a, b) => {
                    let ua = (rewardsSum[a] / counts[a]) + math.sqrt(2 * math.log(t) / counts[a]);
                    let ub = (rewardsSum[b] / counts[b]) + math.sqrt(2 * math.log(t) / counts[b]);
                    return ua >= ub ? a : b;
                });
                const reward = math.random() < means[arm] ? 1 : 0;
                rewardsSum[arm] += reward;
                counts[arm]++;
                total += reward;
            }
            return total;
        },
        'Thompson': () => {
            let successes = { A: 1, B: 1, C: 1 };
            let failures = { A: 1, B: 1, C: 1 };
            let total = 0;
            for (let t = 0; t < budget; t++) {
                const arm = arms.reduce((a, b) => {
                    let sa = randomBeta(successes[a], failures[a]);
                    let sb = randomBeta(successes[b], failures[b]);
                    return sa >= sb ? a : b;
                });
                const reward = math.random() < means[arm] ? 1 : 0;
                if (reward) successes[arm]++; else failures[arm]++;
                total += reward;
            }
            return total;
        }
    };

    let summary = {};
    const optimalReward = 8000;
    for (let [name, func] of Object.entries(sims)) {
        let rewards = [];
        for (let i = 0; i < numRuns; i++) rewards.push(func());
        let avg = rewards.reduce((s, r) => s + r, 0) / numRuns;
        summary[name] = {
            avgReward: avg.toFixed(2),
            regret: (optimalReward - avg).toFixed(2)
        };
    }
    console.log(JSON.stringify(summary, null, 2));
}

simulate();
